src/hazelcore/semantics/Statics.re

```diff
@@ -895,19 +898,40 @@ module Exp = {
     | ListNil(InHole(TypeInconsistent, _))
     | Lam(InHole(TypeInconsistent, _), _, _, _)
     | Inj(InHole(TypeInconsistent, _), _, _)
-    | Case(InHole(TypeInconsistent, _), _, _, _)
+    | Case(InHole(TypeInconsistent, _), _, _)
     | ApPalette(InHole(TypeInconsistent, _), _, _, _) =>
       let operand' = UHExp.set_err_status_operand(NotInHole, operand);
       syn_operand(ctx, operand') |> OptUtil.map(_ => HTyp.Hole);
-    | Var(InHole(WrongLength, _), _, _)
-    | IntLit(InHole(WrongLength, _), _)
-    | FloatLit(InHole(WrongLength, _), _)
-    | BoolLit(InHole(WrongLength, _), _)
-    | ListNil(InHole(WrongLength, _))
-    | Lam(InHole(WrongLength, _), _, _, _)
-    | Inj(InHole(WrongLength, _), _, _)
-    | Case(InHole(WrongLength, _), _, _, _)
-    | ApPalette(InHole(WrongLength, _), _, _, _) => None
+    | Var(InHole(WrongLength | InconsistentBranches(_), _), _, _)
+    | IntLit(InHole(WrongLength | InconsistentBranches(_), _), _)
+    | FloatLit(InHole(WrongLength | InconsistentBranches(_), _), _)
+    | BoolLit(InHole(WrongLength | InconsistentBranches(_), _), _)
+    | ListNil(InHole(WrongLength | InconsistentBranches(_), _))
+    | Lam(InHole(WrongLength | InconsistentBranches(_), _), _, _, _)
+    | Inj(InHole(WrongLength | InconsistentBranches(_), _), _, _)
+    | Case(InHole(WrongLength, _), _, _)
+    | ApPalette(InHole(WrongLength | InconsistentBranches(_), _), _, _, _) =>
+      None
+    | Case(InHole(InconsistentBranches(rule_types), _), scrut, rules) =>
+      switch (syn(ctx, scrut)) {
+      | None => None
+      | Some(pat_ty) =>
+        List.fold_left2(
+          (ty, rule_type, rule) => {
+            switch (ty) {
+            | None => None
+            | Some(_) =>
+              switch (ana_rule(ctx, rule, pat_ty, rule_type)) {
```

Comment:
this should actually be checking that the rule synthesizes the rule_type that the hole reason says it does, which isn't quite the same as analysis

Reply:
(use List.all or similar for this, and then just return HTyp.Hole if that works, rather than folding)